keps/sig-api-machinery/20200309-consistent-resource-versions-semantics.md

```diff
@@ -0,0 +1,216 @@
+	//   ListMeta of the response matches the requested ResourceVersion, and handle the case where it does not. For
+	//   example, the client might fall back to the a request with Limit set.
+	// +optional
+	ResourceVersionMatch ResourceVersionMatch `json:"resourceVersionMatch,omitempty" protobuf:"bytes,10,opt,name=resourceVersionMatch"`
+    ...
+}
+```
+
+### Backward Compatibility
+
+Versions of the kube-apiserver that pre-date the introduction of `ResourceVersionMatch`
+parameter will ignore it. Client will need to tolerate this as documented on ResourceVersionWatch:
+
+- When using ResourceVersionMatchNotOlderThan and Limit is set, clients must handle HTTP 410 “Gone” responses.
+  For example, the client might retry with a newer ResourceVersion or fall back to a ResourceVersion="" request.
+- When using ResourceVersionMatchExact and Limit is unset, clients must verify that the ResourceVersion in the
+  ListMeta of the response matches the requested ResourceVersion, and handle the case where it does not. For
+  example, the client might fall back to the a request with Limit set.
+
+When 'ResourceVersionMatch' is not provided, the behavior is the same as before it was introduced.
+
+### Get support?
+
+`ResourceVersionMatch` can also be added to the get operation for consistency. `Get` currently
+`NotOlderThan` semantics by default and this would add support for `Exact`.
+
+In order to be backward compatible, get responses must include the `ResourceVersion` that the request was served at.
+For list responses this is provided in `ListMeta`, but get responses do not have a wrapper object like `ListMeta`.
+The `ObjectMeta.ResourceVersion` cannot be used because it represents the resource version that the object was
+created or last modified at, not the resource version is was served from.
+
+Options:
+- Don't support `ResourceVersionMatch` for get.
+- Add a header that provides the `ResourceVersion` the get was served at back to clients, e.g. `ServedAtResourceVersion: 43049`
+
+I am currently considering not including `ResourceVersionMatch` since:
+- The header approach sets precidence for an approach I'm not sure we want to encourage in api-machinery
+- get currently does not have the semantic consistency problems of list, and so does not urgently need this parameter
+- it is possible to use list to get a single item at a specific resource version already
+
+### Risks and Mitigations
+
+The main risk to this proposal that it complicates the API surface area, resulting in an API that is
+more difficult to understand and use. But the existing behavior already complicates the API. With
+this proposal, clients at least have the ability to use a consistent set of semantics.
+
+Another risk is that clients will either not realize, or not be sufficiently
+motivated, to update their code to move away from the legacy behavior. This
+can be mitigated a couple ways:
+- Update client bindings (client-go, ...) to discourage using the
```

Comment:
It might be just me with my non-native english, but I had to think what "bindings" is in this context. Maybe "libraries"? or sth like that? [or maybe I still didn't fully understand it?]

Reply:
"client libraries" works here. "client bindings" doesn't really work well here; usually refers to the client code generated from a API spec, which is not really what I'm getting at.